File: drizzle/src/arithmetic.ts
import type { PlanExpressionOperand, Value } from "@cerbos/core";

import { isExpressionOperand, isValueOperand } from "./operands";
import type { ExpressionOperand } from "./operands";

/**
 * Constant folding in JavaScript's IEEE double space — CEL's own number space — for the
 * arithmetic and comparisons SQL cannot be trusted with: division by zero, NaN and infinities.
 */

export const ARITHMETIC_OPERATORS: Record<string, string> = {
  add: "+",
  sub: "-",
  mult: "*",
  div: "/",
  mod: "%",
};

export type LeafComparisonOperator = "eq" | "ne" | "lt" | "le" | "gt" | "ge";

/**
 * The arithmetic that folds exactly in IEEE doubles. CEL's `%` is integer-only while Cerbos
 * attribute values are always doubles, so a modulus is a no-overload error that denies every row
 * at check time; folding it with JavaScript's `%` would answer a question CEL refused.
 */
type Fold = (left: number, right: number) => number;

const FOLDABLE_ARITHMETIC = new Map<string, Fold>([
  ["add", (left, right) => left + right],
  ["sub", (left, right) => left - right],
  ["mult", (left, right) => left * right],
  ["div", (left, right) => left / right],
]);

/** The foldable binary arithmetic node `operand` is, if it is one. */
const foldableArithmetic = (
  operand: PlanExpressionOperand,
):
  | { apply: Fold; left: PlanExpressionOperand; right: PlanExpressionOperand }
  | undefined => {
  if (!isExpressionOperand(operand) || operand.operands.length !== 2) {
    return undefined;
  }
  const apply = FOLDABLE_ARITHMETIC.get(operand.operator);
  const [left, right] = operand.operands;
  return apply && left && right ? { apply, left, right } : undefined;
};

/**
 * The value of an operand that is a number literal, or `+ - * /` over number literals. Anything
 * else — a column, a string, a modulus — is `undefined`.
 */
export const resolveConstantNumber = (
  operand: PlanExpressionOperand,
): number | undefined => {
  if (isValueOperand(operand)) {
    return typeof operand.value === "number" ? operand.value : undefined;
  }
  const arithmetic = foldableArithmetic(operand);
  if (!arithmetic) {
    return undefined;
  }
  const left = resolveConstantNumber(arithmetic.left);
  const right = resolveConstantNumber(arithmetic.right);
  return left === undefined || right === undefined
    ? undefined
    : arithmetic.apply(left, right);
};

/**
 * A division whose denominator is not a known non-zero constant, so the row-level result
 * may be a CEL NaN or signed infinity that SQL cannot represent.
 */
const isZeroCapableDivision = (
  operand: PlanExpressionOperand,
): operand is ExpressionOperand => {
  if (!isExpressionOperand(operand) || operand.operator !== "div") {
    return false;
  }
  // A division of two constants already folds to an exact IEEE value, and the
  // constant/NaN paths render it more tightly than a CASE can.
  if (resolveConstantNumber(operand) !== undefined) {
    return false;
  }
  const denominatorOperand = operand.operands[1];
  if (!denominatorOperand) {
    return false;
  }
  const denominator = resolveConstantNumber(denominatorOperand);
  return denominator === undefined || denominator === 0;
};

/**
 * The first zero-capable division anywhere in an arithmetic tree. The comparison need not
 * sit directly on the division: `div(a, a) + 1.0 != 2.0` composes addition on top of a
 * result that can be NaN, and lowering that to `NULL + 1` excludes the one row CEL allows.
 */
export const findZeroCapableDivision = (
  operand: PlanExpressionOperand,
): ExpressionOperand | undefined => {
  if (isZeroCapableDivision(operand)) {
    return operand;
  }
  if (
    !isExpressionOperand(operand) ||
    !(operand.operator in ARITHMETIC_OPERATORS)
  ) {
    return undefined;
  }
  for (const child of operand.operands) {
    const found = findZeroCapableDivision(child);
    if (found) {
      return found;
    }
  }
  return undefined;
};

/**
 * Evaluate an arithmetic tree in JavaScript's IEEE double space with `target` replaced by
 * `substitute`, so a NaN or infinity produced by a division propagates through the
 * surrounding arithmetic exactly as CEL propagates it. Returns undefined when any other
 * leaf is not a constant — SQL cannot carry a non-finite value alongside a column, and
 * guessing would return rows the PDP denies.
 */
export const foldWithSubstitution = (
  operand: PlanExpressionOperand,
  target: PlanExpressionOperand,
  substitute: number,
): number | undefined => {
  if (operand === target) {
    return substitute;
  }
  if (isValueOperand(operand)) {
    return typeof operand.value === "number" ? operand.value : undefined;
  }
  // A modulus is not foldable, so it fails closed here rather than being answered.
  const arithmetic = foldableArithmetic(operand);
  if (!arithmetic) {
    return undefined;
  }
  const left = foldWithSubstitution(arithmetic.left, target, substitute);
  const right = foldWithSubstitution(arithmetic.right, target, substitute);
  return left === undefined || right === undefined
    ? undefined
    : arithmetic.apply(left, right);
};

export const evaluateConstantNumberComparison = (
  operator: LeafComparisonOperator,
  left: number,
  right: number,
): boolean => {
  if (
    operator !== "eq" &&
    operator !== "ne" &&
    (Number.isNaN(left) || Number.isNaN(right))
  )
    return false;
  switch (operator) {
    case "eq":
      return left === right;
    case "ne":
      return left !== right;
    case "lt":
      return left < right;
    case "le":
      return left <= right;
    case "gt":
      return left > right;
    case "ge":
      return left >= right;
  }
};

/** A comparison between two literals of the same CEL type; `undefined` when CEL has no overload. */
export const evaluateScalarValueComparison = (
  operator: LeafComparisonOperator,
  left: Value,
  right: Value,
): boolean | undefined => {
  const bothNumbers = typeof left === "number" && typeof right === "number";
  const bothStrings = typeof left === "string" && typeof right === "string";
  const bothBooleans = typeof left === "boolean" && typeof right === "boolean";
  const bothNull = left === null && right === null;
  if (!bothNumbers && !bothStrings && !bothBooleans && !bothNull) {
    return undefined;
  }
  switch (operator) {
    case "eq":
      return left === right;
    case "ne":
      return left !== right;
    case "lt":
      return bothNumbers || bothStrings ? left < right : undefined;
    case "le":
      return bothNumbers || bothStrings ? left <= right : undefined;
    case "gt":
      return bothNumbers || bothStrings ? left > right : undefined;
    case "ge":
      return bothNumbers || bothStrings ? left >= right : undefined;
  }
};

File: sequel/lib/cerbos/sequel/translator/arithmetic.rb
# frozen_string_literal: true

module Cerbos
  module Sequel
    class Translator
      # +add+, +sub+, +mult+, +mod+ and +div+.
      #
      # CEL arithmetic on attributes uses doubles, so a division by zero is not an error: it
      # gives NaN or an Infinity. SQL cannot hold those values. The translator keeps them as
      # {Values::IEEEConstant} and {Values::ConditionalValue} until the comparison around them
      # calculates each branch.
      module Arithmetic
        private

        def arithmetic(operator, left, right)
          # A division that can give a value which is not finite stays as branches until a
          # comparison calculates it. More arithmetic on those branches has no SQL equivalent,
          # so the adapter raises instead of making an incorrect filter.
          require_scalars(operator, left, right)
          reject_int_beside_non_int(operator, left, right)

          # CEL uses `+` for strings and for numbers. SQL does not. On SQLite and MySQL,
          # `'a' + 'b'` is an addition of numbers, and it changes both sides into 0. Thus string
          # operands need the concatenation operation of the dialect.
          if operator == "add" && (string_valued?(left) || string_valued?(right))
            return left + right if left.is_a?(::String) && right.is_a?(::String)

            return dialect.concat(left, right)
          end

          if operator == "mod" && !(cel_int?(left) && cel_int?(right))
            raise UnsupportedOperatorError,
              "% has no double overload in CEL, and every number in a request attribute is a " \
              "double, so % over an attribute that has not gone through int() is an error that " \
              "denies the row. SQL computes a remainder instead. Wrap the operands in int()."
          end

          if left.is_a?(Numeric) && right.is_a?(Numeric)
            return left.public_send(ARITHMETIC.fetch(operator), right)
          end

          if cel_int?(left) && cel_int?(right)
            # CEL's `%` by zero is an error, which denies the row under either polarity.
            # PostgreSQL raises instead, failing the whole query; NULLIF makes it UNKNOWN, as
            # SQLite and MySQL already make it.
            right = SqlSupport.function("NULLIF", [right, 0]) if operator == "mod" && !right.is_a?(Numeric)
            return record_cel_type(SqlSupport.infix(ARITHMETIC.fetch(operator), left, right), :int)
          end

          # CEL holds every attribute number as a double. PostgreSQL and MySQL would compute an
          # integer or decimal column with a literal like 0.1 in exact decimal, so
          # `aNumber * 0.1 == 0.3` would hold for 3 where CEL computes 0.30000000000000004.
          left, right = [left, right].map { |operand| exact_numeric_column?(operand) ? as_double(operand) : operand }
          record_cel_type(SqlSupport.infix(ARITHMETIC.fetch(operator), left, right), :double)
        end

        def exact_numeric_column?(value)
          SqlSupport.sql_node?(value) && EXACT_NUMERIC_COLUMN_TYPES.include?(column_type(value))
        end

        # CEL has no overload mixing an int with a double: `int(x) + R.attr.d` is an error that
        # denies the row under either polarity, where SQL adds the two numbers and a negation
        # turns the sum into a grant. An int() result beside an operand that is not certainly an
        # int (a column, whose attribute is a double, or a fractional constant) is refused.
        def reject_int_beside_non_int(operator, left, right)
          mixed = (cel_type(left) == :int && !cel_int?(right)) ||
            (cel_type(right) == :int && !cel_int?(left))
          return unless mixed

          raise UnsupportedOperatorError,
            "#{operator} of an int() result and an operand that is not an int: CEL has no " \
            "overload mixing int and double, so the expression is an error that denies the row, " \
            "but SQL computes it. Every number in a request attribute is a double; wrap both " \
            "operands in int(), or neither."
        end

        # An operand CEL holds as an int: an int() result, or arithmetic on those. A whole
        # constant counts too, since the plan does not say whether a literal was `2` or `2.0`
        # and CEL's type checker rejects an int mixed with a double.
        def cel_int?(value)
          return value.finite? && value == value.truncate if value.is_a?(Float)
          return true if value.is_a?(Integer)

          cel_type(value) == :int
        end

        # Divides as doubles, like CEL. Otherwise SQLite and PostgreSQL make `5 / 2` into `2`.
        # Two ints are the exception: CEL's int division truncates toward zero.
        def divide(numerator, denominator)
          require_scalars("div", numerator, denominator)
          reject_int_beside_non_int("div", numerator, denominator)
          return int_divide(numerator, denominator) if int_division?(numerator, denominator)

          if numerator.is_a?(Numeric) && denominator.is_a?(Numeric)
            return divide_constants(numerator.to_f, denominator.to_f)
          end

          # A constant denominator that is not zero can never divide by zero. Thus a plain
          # division is exact, and it keeps the SQL small.
          if denominator.is_a?(Numeric) && !denominator.to_f.zero?
            return record_cel_type(SqlSupport.infix("/", as_double(numerator), denominator.to_f), :double)
          end

          divide_with_zero_denominator(numerator, denominator)
        end

        # Int division: both operands are CEL ints and at least one is an int() result (or int
        # arithmetic on one), since a bare whole constant may have been written `2.0`.
        def int_division?(numerator, denominator)
          (cel_type(numerator) == :int || cel_type(denominator) == :int) &&
            cel_int?(numerator) && cel_int?(denominator)
        end

        # CEL's `int / int` truncates toward zero, as SQLite's and PostgreSQL's integer `/`
        # and MySQL's `DIV` do. A zero divisor is a CEL error that denies the row under either
        # polarity, where PostgreSQL aborts the query, so only a non-zero constant divisor is
        # translated.
        def int_divide(numerator, denominator)
          unless denominator.is_a?(Numeric) && !denominator.zero?
            raise UnsupportedOperatorError,
              "int division by a value that may be zero: CEL makes an error that denies the " \
              "row, but PostgreSQL aborts the whole query, so only a non-zero constant divisor " \
              "is translated"
          end

          return (numerator.to_i.to_r / denominator.to_i).truncate if numerator.is_a?(Numeric)

          record_cel_type(dialect.int_divide(numerator, denominator.to_i), :int)
        end

        def divide_constants(numerator, denominator)
          return Values::IEEEConstant.new(value: numerator / denominator) if denominator.zero?

          numerator / denominator
        end

        # A division by zero is not an error in CEL, because CEL arithmetic on attributes uses
        # doubles. IEEE-754 gives NaN for 0/0, +Infinity for a positive numerator, and -Infinity
        # for a negative one. SQL cannot hold those three values, and NULL is not equal to any of
        # them: `NaN != 1.0` is TRUE in CEL, but `NULL != 1.0` is UNKNOWN in SQL, and thus a
        # NULL would remove a row that the PDP permits.
        #
        # The translator keeps the three cases as branches. The comparison around the division
        # then calculates each branch, in the same way as any other constant that is not finite.
        def divide_with_zero_denominator(numerator, denominator)
          # The denominator is a constant zero, so the sign of that zero is known.
          return zero_denominator_value(numerator, zero_sign(denominator)) if denominator.is_a?(Numeric)

          # The denominator is row-dependent. SQL cannot tell -0.0 from 0.0 — both satisfy
          # `= 0` and no portable function reads the sign bit — so the sign of an Infinity is
          # unknowable here. The one shape that stays safe is a division of a value by itself:
          # the denominator can only be zero when the numerator is zero too, which gives NaN,
          # and NaN has no sign question.
          unless numerator == denominator
            raise UnsupportedOperatorError,
              "Cannot divide by a column that may be zero: IEEE-754 keeps the sign of a zero, " \
              "SQL cannot tell -0.0 from 0.0, and thus the sign of the Infinity is unknown. " \
              "Divide by a constant, or keep this shape out of the policy."
          end

          Values::ConditionalValue.new(
            condition: SqlSupport.comparison("eq", as_double(denominator), 0.0),
            then_value: Values::IEEEConstant.new(value: Float::NAN),
            else_value: SqlSupport.infix("/", as_double(numerator), as_double(denominator))
          )
        end

        # IEEE-754 keeps the sign of a zero. `2.0 / -0.0` is -Infinity, not +Infinity, because the
        # sign of the result is the sign of the numerator against the sign of the denominator.
        def zero_sign(denominator)
          (1.0 / denominator.to_f).negative? ? -1.0 : 1.0
        end

        def zero_denominator_value(numerator, sign)
          if numerator.is_a?(Numeric)
            return Values::IEEEConstant.new(value: numerator.to_f / (sign * 0.0))
          end

          Values::ConditionalValue.new(
            condition: SqlSupport.comparison("eq", as_double(numerator), 0.0),
            then_value: Values::IEEEConstant.new(value: Float::NAN),
            else_value: Values::ConditionalValue.new(
              condition: SqlSupport.comparison("gt", as_double(numerator), 0.0),
              then_value: Values::IEEEConstant.new(value: sign * Float::INFINITY),
              else_value: Values::IEEEConstant.new(value: -sign * Float::INFINITY)
            )
          )
        end
      end
    end
  end
end
